### PR TITLE
CRM-19100: 4.7.8 SQL Incremental Update May Fail for UTF8

### PR DESCRIPTION
**Pull request by jitendrapurohit**
- [CRM-19100: 4.7.8 SQL Incremental Update May Fail for UTF8](https://issues.civicrm.org/jira/browse/CRM-19100)

### CODE REVIEW

File: CRM/Upgrade/Incremental/php/FourSeven.php [hunk 1]
```diff
@@ -726,4 +727,19 @@ public function upgradeMailingFKs(CRM_Queue_TaskContext $ctx) {
     return TRUE;
   }
 
+  /**
+   * CRM-19100 - Alter Index and Type for Image URL
+   * @return bool
+   */
+  public static function alterIndexAndTypeForImageURL() {
+    $dao = CRM_Core_DAO::executeQuery("SHOW INDEX FROM civicrm_contact WHERE KEY_NAME = 'index_image_url'");
```
Comment (seamuslee001) [inline CRM/Upgrade/Incremental/php/FourSeven.php]: suggest using this function instead https://github.com/civicrm/civicrm-core/blob/master/CRM/Core/BAO/SchemaHandler.php#L571

File: CRM/Upgrade/Incremental/php/FourSeven.php [hunk 2]
```diff
@@ -726,4 +727,19 @@ public function upgradeMailingFKs(CRM_Queue_TaskContext $ctx) {
...
+    if ($dao->N) {
+      CRM_Core_DAO::executeQuery("ALTER TABLE `civicrm_contact` DROP INDEX `index_image_url`");
+    }
+    CRM_Core_DAO::executeQuery("ALTER TABLE `civicrm_contact` CHANGE `image_URL` `image_URL` TEXT CHARACTER SET utf8 COLLATE utf8_unicode_ci NULL DEFAULT NULL COMMENT 'optional URL for preferred image (photo, logo, etc.) to display for this contact.'");
+    CRM_Core_DAO::executeQuery("CREATE INDEX `index_image_url` ON `civicrm_contact` ( image_URL(128) )");
```
Comment (seamuslee001) [inline CRM/Upgrade/Incremental/php/FourSeven.php]: maybe use this function as well https://github.com/civicrm/civicrm-core/blob/master/CRM/Core/BAO/SchemaHandler.php#L418

File: CRM/Upgrade/Incremental/php/FourSeven.php (hunk 1 above)
Comment (jitendrapurohit) [inline CRM/Upgrade/Incremental/php/FourSeven.php]: Done :-)

**Comment by monishdeb** [comment]
Jenkin test this please

**Comment by JoeMurray** [comment]
I would have preferred to see all of this being done just in SQL template rather than in PHP. But I think it is all good and ready to merge.

**Comment by monishdeb** [comment]
@JoeMurray ya we tried earlier but then on some scenario there was no image_url index present, reason why we cannot directly drop that index without ensuring its presence otherwise it will throw
 mysql error ```cannot drop index, unable to find column/key```. And already we got a DAO function ```dropIndexIfExists()```, so the code was then moved to php script.

**Comment by PalanteJon** [comment]
I'm with @monishdeb on this one - [CRM-18526](https://issues.civicrm.org/jira/browse/CRM-18526) is a good example of what happens when a SQL-only upgrade script attempts to edit a SQL index!

**Comment by JoeMurray** [comment]
Agreed then. Quite concerning to me that the index is sometimes not there...why not? But that's a different can of worms.

**Pull request merged**